File: sdk/agentflow/exceptions.py
class AgentFlowError(Exception):
    pass


class AuthError(AgentFlowError):
    pass


class PermissionDeniedError(AgentFlowError):
    pass


class RateLimitError(AgentFlowError):
    def __init__(self, message: str, retry_after: int = 0):
        super().__init__(message)
        self.retry_after = retry_after


class DataFreshnessError(AgentFlowError):
    pass


class EntityNotFoundError(AgentFlowError):
    def __init__(self, entity_type: str, entity_id: str, message: str | None = None):
        super().__init__(message or f"{entity_type}/{entity_id} not found")
        self.entity_type = entity_type
        self.entity_id = entity_id

File: sdk-ts/src/exceptions.ts
export class AgentFlowError extends Error {
  readonly statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.name = "AgentFlowError";
    this.statusCode = statusCode;
  }
}

export class AuthError extends AgentFlowError {
  constructor(message = "Unauthorized") {
    super(message, 401);
    this.name = "AuthError";
  }
}

export class PermissionDeniedError extends AgentFlowError {
  constructor(message = "Forbidden") {
    super(message, 403);
    this.name = "PermissionDeniedError";
  }
}

export class RateLimitError extends AgentFlowError {
  readonly retryAfter: number;

  constructor(message = "Rate limit exceeded", retryAfter = 0) {
    super(message, 429);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

export class DataFreshnessError extends AgentFlowError {
  constructor(message: string) {
    super(message);
    this.name = "DataFreshnessError";
  }
}

export class EntityNotFoundError extends AgentFlowError {
  readonly entityType: string;
  readonly entityId: string;

  constructor(entityType: string, entityId: string, message?: string) {
    super(message ?? `${entityType}/${entityId} not found`, 404);
    this.name = "EntityNotFoundError";
    this.entityType = entityType;
    this.entityId = entityId;
  }
}
